lib/base/request.js

```diff
@@ -365,20 +366,34 @@ class Request extends EventEmitter {
   }
 
   /**
-   * Sets request to `stream` mode and pulls all rows from all recordsets to a given stream.
+   * Wrap original request in a Readable stream that supports back pressure and pipe to the Writable stream.
+   * It also sets request to `stream` mode and pulls all rows from all recordsets to a given stream.
    *
-   * @param {Stream} stream Stream to pipe data into.
+   * @param {Stream} writableStream Stream to pipe data into.
    * @return {Stream}
    */
-
-  pipe (stream) {
-    this.stream = true
-    this.on('row', stream.write.bind(stream))
-    this.on('done', () => {
-      setImmediate(() => stream.end())
+  pipe (writableStream) {
+    const origRequest = this
+    origRequest.stream = true
+    origRequest.pause()
+    const readableStream = new Readable({
+      objectMode: true,
+      read (/* size */) {
+        origRequest.resume()
+      }
+    })
+    origRequest.on('row', (row) => {
+      if (!readableStream.push(row)) {
+        origRequest.pause()
+      }
+    })
+    origRequest.on('error', (error) => {
+      readableStream.emit('error', error)
+    })
+    origRequest.on('done', () => {
+      readableStream.push(null)
     })
-    stream.emit('pipe', this)
```

Comment:
I'm slightly concerned that we are losing the pipe event here. Will the pipe event fire on `readableStream` in the next tick - can a user still bind to a `pipe` event?

Reply:
We can still emit the pipe event so that it isn't lost.

Reply:
However I think that maybe it is already being emitted from the ReadableStream to the WritableStream. Maybe you are suggesting it is emitted on the original Request?